File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        
        Map <String, Integer> player_ = new HashMap<>();
        Map <Integer, String> rank_ = new HashMap<>();
        for(int i = 0; i < players.length; i++) {
            player_.put(players[i], i);
            rank_.put(i, players[i]);
        }
        
        for(String i : callings) {
            int cur_rank = player_.get(i);
            int fro_rank = cur_rank - 1;
            String fro_player = rank_.get(fro_rank);
            
            player_.put(fro_player, cur_rank);
            player_.put(i, fro_rank);
            
            rank_.put(fro_rank, i);
            rank_.put(cur_rank, fro_player);
        }
        String[] answer = new String[players.length];
        int cnt = 0;
        for (String i : rank_.values()){
            answer[cnt++] = i;
        }
        
        return answer;
    }
}

File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    player_ = {rank:player for rank, player in enumerate(players)}
    rank_ = {player:rank for rank, player in enumerate(players)}
    
    for i in callings:
        cur_rank = rank_[i]
        fro_rank = cur_rank - 1
        fro_player = player_[fro_rank]
        
        # player_ 딕셔너리 정렬
        player_[cur_rank] = fro_player
        player_[fro_rank] = i
        
        # rank_ 딕셔너리 정렬
        rank_[i] = fro_rank
        rank_[fro_player] = cur_rank

    return list(player_.values())